Source/Dafny/Resolver.cs

```diff
@@ -15329,6 +15437,67 @@ public class ResolveOpts {
       }
     }
 
+    private void AssumeCompilableTypeForBoundedVariables(ComprehensionExpr e) {
+      foreach (BoundVar v in e.AllBoundVars) {
+        v.AssumeCompilableTypeIfAny(); // Reinstate the runtime testable type since verification uses bounded variable
+      }
+    }
+
+    private void AssumeOriginalTypeForBoundedVariables(ComprehensionExpr e) {
+      foreach (BoundVar v in e.AllBoundVars) {
+        v.AssumeOriginalType(); // Reinstate the runtime testable type since verification uses bounded variable
+      }
+    }
+
+    private void WithoutReporting(System.Action callback) {
+      var savedReporter = reporter;
+      reporter = new ErrorReporterSink(); // Prevent errors from being duplicated
+      callback();
+      reporter = savedReporter;
+    }
+
+    private void ScopePushBoundVarsWithoutAssumptions(ComprehensionExpr e) {
+      // For the term only, we can assume the inferred type which has to be proved later
+      foreach (BoundVar v in e.AllBoundVars) {
+        // Previously, the type has to be runtime-testable for the range at least.
+        // Now the verifier will guarantee that the type satisfies all the subset type constraints, if any.
+        // Reinstate the original type set by the user, if any
+        v.AssumeOriginalType();
+
+        ScopePushAndReport(scope, v, "bound-variable");
+        if (v.Type is InferredTypeProxy inferredProxy) {
+          Contract.Assert(!inferredProxy.KeepConstraints); // in general, this proxy is inferred to be a base type
+        }
+      }
+    }
+
+    /// Returns the first non-runtime testable type of the comprehension if it exists
+    /// Ensures ret != null ==> !ret.IsCompilable() && ret is one of the type of e.AllBoundVars[i]
+    private void ScopePushBoundVarsAssumingCompilable(ComprehensionExpr e, ResolveOpts opts, [CanBeNull] ResolveTypeOption resolveTypeOption = null, bool typeQuantifier = false) {
+      if (resolveTypeOption == null) {
+        resolveTypeOption = new ResolveTypeOption(ResolveTypeOptionEnum.InferTypeProxies);
+      }
+      foreach (BoundVar v in e.AllBoundVars) {
+        var typeArgs = e is QuantifierExpr q ? (typeQuantifier ? q.TypeArgs : null) : null;
+        ResolveType(v.tok, v.Type, opts.codeContext, resolveTypeOption, typeArgs);
+        // If the type can be tested at run time, we keep the same scope
+        // Else, the the scoped type is the upper type that can be tested.
+        if (!v.Type.IsCompilable() && !opts.isSpecification && !opts.codeContext.IsGhost) {
+          var collectionVarType = v.Type is InferredTypeProxy ? new InferredTypeProxy() : v.Type.GetCompilableParentType(); ;
+          if (v.Type is InferredTypeProxy) {
+            AddXConstraint(v.tok, "SubsetTypeOfCompilable", v.Type, collectionVarType,
+              $"Type of variable {v.DisplayName} in range is '{{1}}', the range should prove it's of type '{{0}}' (requested type), but that's not the case."
```

Comment:
New formulation:
"Type of variable {v.DisplayName} was requested to be of type {{0}}, but it can only be of type {{1}} based on the range"

Note that I am not able to test this error, as this constraint is very permissive and mostly trigger other constraints.